source/heating_model/compositional_heating.cc

```diff
@@ -0,0 +1,172 @@
+      double sum_composition = 0.0;
+
+      std::vector<double> x_comp = compositional_fields;
+      for ( unsigned int i=0; i < x_comp.size(); ++i)
+        {
+          if (field_used_in_heat_production_averaging[i] == 1)
+            {
+              // clip the compositional fields so they are between zero and one
+              // and sum them for normalization purposes
+              x_comp[i] = std::min(std::max(x_comp[i], 0.0), 1.0);
+              sum_composition += x_comp[i];
+            }
+          else
+            x_comp[i] = 0.0;
+        }
+
+      if (sum_composition >= 1.0)
+        {
+          volume_fractions[0] = 0.0;  //background material
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1]/sum_composition;
+        }
+      else
+        {
+          volume_fractions[0] = 1.0 - sum_composition; //background material
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1];
+        }
+      return volume_fractions;
+    }
+
+
+    template <int dim>
+    void
+    CompositionalHeating<dim>::
+    evaluate (const MaterialModel::MaterialModelInputs<dim> &material_model_inputs,
+              const MaterialModel::MaterialModelOutputs<dim> &material_model_outputs,
+              HeatingModel::HeatingModelOutputs &heating_model_outputs) const
+    {
+
+      for (unsigned int q = 0; q < heating_model_outputs.heating_source_terms.size(); ++q)
+        {
+
+          // Compute compositional volume fractions
+          const std::vector<double> volume_fractions = compute_volume_fractions(material_model_inputs.composition[q]);
+
+          // Calculate average compositional heat production
+          double compositional_heat_production = 0.;
+          for (unsigned int c=0; c< volume_fractions.size(); ++c)
+            {
+              compositional_heat_production += volume_fractions[c] * heating_values[c];
+            }
+
+          heating_model_outputs.heating_source_terms[q] = compositional_heat_production;
+
+          heating_model_outputs.lhs_latent_heat_terms[q] = 0.0;
+
+        }
+    }
+
+
+    template <int dim>
+    void
+    CompositionalHeating<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Heating model");
+      {
+        prm.enter_subsection("Compositional heating");
+        {
+          prm.declare_entry("Compositional heating values","0",
+                            Patterns::List (Patterns::Double(0)),
+                            "List of heat production per unit volume values for "
+                            "background and compositional fields, for a total of "
+                            "N+1 values, where N is the number of compositional fields. "
+                            "Units: W/m3.");
+          prm.declare_entry ("Use compositional field for heat production averaging", "1",
+                             Patterns::List(Patterns::Integer(0,1)),
```

Comment:
Would a `Patterns::List(Patterns::Bool)` work? Then the input would look like `true,false,false`. If you prefer 0 and 1 in the input file keeping the integer here would be fine with me as well, I would just like to store it internally as bool to make its purpose clear.